samza-kafka/src/main/scala/org/apache/samza/checkpoint/kafka/KafkaCheckpointManager.scala

```diff
@@ -19,258 +19,212 @@
-      info("Already existing checkpoint mapping.  Merging new offsets")
-      taskNamesToOffsets ++= readCheckpointsFromLog()
+      info("Updating existing checkpoint mappings")
+      taskNamesToOffsets ++= readCheckpoints()
+      info("Updating existing checkpoint mappings" + taskNamesToOffsets)
     }
 
     val checkpoint = taskNamesToOffsets.get(taskName).getOrElse(null)
 
-    info("Got checkpoint state for taskName %s: %s" format(taskName, checkpoint))
-
+    info(s"Got checkpoint state for taskName $taskName $checkpoint")
     checkpoint
   }
 
   /**
-   * Read through entire log, discarding changelog mapping, and building map of TaskNames to Checkpoints
-   */
-  def readCheckpointsFromLog(): Map[TaskName, Checkpoint] = {
-    val checkpoints = mutable.Map[TaskName, Checkpoint]()
+    * @inheritdoc
+    */
+  override def writeCheckpoint(taskName: TaskName, checkpoint: Checkpoint) {
+    val key = new KafkaCheckpointLogKey(grouperFactory, taskName, KafkaCheckpointLogKey.CHECKPOINT_TYPE)
+    val keyBytes = checkpointKeySerde.toBytes(key)
+    val msgBytes = checkpointMsgSerde.toBytes(checkpoint)
+    val envelope = new OutgoingMessageEnvelope(checkpointSsp, keyBytes, msgBytes)
+    val retryBackoff: ExponentialSleepStrategy = new ExponentialSleepStrategy
 
-    def shouldHandleEntry(key: KafkaCheckpointLogKey) = key.isCheckpointKey
+    retryBackoff.run(
+      loop => {
+        systemProducer.send(taskName.getTaskName, envelope)
+        systemProducer.flush(taskName.getTaskName) // make sure it is written
+        debug("Completed writing checkpoint=%s into %s topic for system %s." format(checkpoint, checkpointTopic, checkpointSystem))
+        loop.done
+      },
 
-    def handleCheckpoint(payload: ByteBuffer, checkpointKey:KafkaCheckpointLogKey): Unit = {
-      val taskName = checkpointKey.getCheckpointTaskName
-      val checkpoint = serde.fromBytes(Utils.readBytes(payload))
-      checkpoints.put(taskName, checkpoint) // replacing any existing, older checkpoints as we go
-    }
+      (exception, loop) => {
+        warn(s"Failed to write checkpoint. key: $key", exception)
+      }
+    )
+  }
 
-    readLog(shouldHandleEntry, handleCheckpoint)
-    checkpoints.toMap /* of the immutable kind */
+  /**
+    * @inheritdoc
+    */
+  override def clearCheckpoints = {
+    info("Clear checkpoint stream %s in system %s" format(checkpointTopic, checkpointSystem))
+    systemAdmin.clearStream(checkpointSpec)
   }
 
-  private def getSSPMetadata(topic: String, partition: Partition): SystemStreamPartitionMetadata = {
-    val metaDataMap: java.util.Map[String, SystemStreamMetadata] = systemAdmin.getSystemStreamMetadata(Collections.singleton(topic))
-    val checkpointMetadata: SystemStreamMetadata = metaDataMap.get(topic)
-    if (checkpointMetadata == null) {
-      throw new SamzaException("Cannot get metadata for system=%s, topic=%s" format(systemName, topic))
+  override def stop = {
+    if (systemProducer != null) {
```

Comment:
We already throw an exception if checkpointMetadata is null... Do you suggest also logging an error?

Reply:
I was talking about the systemProducer != null and systemConsumer != null in the stop method. If they are null, that's unexpected and should be logged as an error so that we know this happens.